File: CSharp/Studies/Studies/Arrays/MaxProfitExercises.cs
using System;

namespace Studies.Arrays
{
    public static class MaxProfitExercises
    {
        /// <summary>
        /// https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/564/
        /// </summary>
        /// <param name="prices">Prices of various stocks indexed by day</param>
        /// <returns>The max profit derived from buying and selling stock from <paramref name="prices"/></returns>
        public static int MaxProfit(int[] prices)
        {
            // for an array [a, b, c, d]
            // (b - a) + (d - c) > (d - a)
            // only when b > c

            var buying = false;
            var profit = 0;
            var purchased = 0;
            var maxProfit = 0;

            for (var i = 0; i < prices.Length - 1; i++)
            {
                var potentialProfit = prices[i + 1] - prices[i];

                if (potentialProfit > 0)
                {
                    purchased = buying ? purchased : prices[i];
                    profit = prices[i + 1] - purchased;
                    buying = true;
                }
                else
                {
                    maxProfit += profit;
                    buying = false;
                }
            }

            if (buying) maxProfit += profit;

            return maxProfit;
        }

        /// <summary>
        /// A recursive version
        /// https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/564/
        /// </summary>
        /// <param name="prices">Prices of various stocks indexed by day</param>
        /// <returns>The max profit derived from buying and selling stock from <paramref name="prices"/></returns>
        public static int MaxProfit2(int[] prices)
        {
            return MaxProfit2(prices, 0, 0);
        }

        private static int MaxProfit2(int[] prices, int nextIndex, int maxProfit)
        {
            var purchase = FindPurchaseIndex(prices, nextIndex);
            
            if (!purchase.Found)
            {
                return maxProfit;
            }
            
            var sell = FindSellIndex(prices, purchase.Index + 1);
            maxProfit += prices[sell.Index] - prices[purchase.Index];
                
            return MaxProfit2(prices, sell.Index + 1, maxProfit);
        }

        private static FindResult FindPurchaseIndex(int[] prices, int nextIndex) 
        {
            if ((nextIndex + 1) >= prices.Length)
            {
                return new FindResult
                {
                    Found = false,
                    Index = nextIndex
                };
            }

            if (prices[nextIndex + 1] > prices[nextIndex])
            {
                return new FindResult 
                {
                    Found = true,
                    Index = nextIndex
                };
            }

            nextIndex++;
            return FindPurchaseIndex(prices, nextIndex);
        }

        private static FindResult FindSellIndex(int[] prices, int nextIndex)
        {
            if ((nextIndex + 1) >= prices.Length || prices[nextIndex + 1] < prices[nextIndex])
            {
                return new FindResult
                {
                    Found = true,
                    Index = nextIndex
                };
            }

            nextIndex++;
            return FindSellIndex(prices, nextIndex);
        }

        private class FindResult
        {
            public bool Found { get; set; }

            public int Index { get; set; }
        }
    }
}

File: js/lib/arrays/MaxProfitExercises.js
class MaxProfitExercises {
    
    /**
     * 
     * @param {Array} prices - Prices of various stocks indexed by day
     * @summary https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/564/
     * 
     * @returns The max profit derived from buying and selling stock from {prices}
     */
    static maxProfit(prices) {
        if (prices === undefined || !Array.isArray(prices)) return 0;

        let buying = false;
        let profit = 0;
        let purchased = 0;
        let maxProfit = 0; 

        for(let i = 0; i < prices.length - 1; i++) {
            var potentialProfit = prices[i + 1] - prices[i];

            if (potentialProfit > 0) {
                purchased = buying ? purchased : prices[i];
                profit = prices[i + 1] - purchased;
                buying = true;
            } else {
                maxProfit += profit;
                buying = false;
            }
        }

        if (buying) maxProfit += profit;

        return maxProfit;
    }

    /**
     * 
     * @param {Array} prices - Prices of various stocks indexed by day
     * @summary https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/564/
     * 
     * @returns The max profit derived from buying and selling stock from {prices}
     */
    static maxProfit2(prices) {
        return this.maxProfitRecursion(prices, 0, 0);
    }

    static maxProfitRecursion(prices, idx, maxProfit) {
        let purchase = this.findPurchaseIndex(prices, idx);

        if (!purchase.found) return maxProfit;

        let sell = this.findSellIndex(prices, idx + 1);
        maxProfit += prices[sell.index] - prices[purchase.index];

        return this.maxProfitRecursion(prices, sell.index + 1, maxProfit);
    }

    static findPurchaseIndex(prices, idx) {
        if (idx >= (prices.length - 1)) return new FindResult(false, -1);

        if ((prices[idx + 1] - prices[idx]) > 0) return new FindResult(true, idx);

        return this.findPurchaseIndex(prices, ++idx);
    }

    static findSellIndex(prices, idx) {
        if (idx >= (prices.length - 1) || (prices[idx] - prices[idx + 1]) > 0) return new FindResult(true, idx);

        return this.findSellIndex(prices, ++idx);
    }
}

class FindResult {
    constructor(found, idx) {
        this.found = found;
        this.index = idx;
    }       
}

module.exports = MaxProfitExercises;